### ISSUE TITLE
Support for vendor locations (pam_shells and /etc/shells)

### ISSUE THREAD

**Issue by dvzrv** [linux-pam/linux-pam#498]
Hi! I'm currently looking into the integration of /etc/shells in the packaging ecosystem on Arch Linux.
Our current way of dealing with /etc/shells is as follows:

* some shells (e.g. sh and bash) are in the file by default (probably for historical reasons)
* other shells sed themselves in and out of the file in package addition/ removal hooks

This means, that /etc/shells provides a form of state, which needs to be maintained by the package management system or the user, or by other means.
When looking at the concept of immutable systems with vendor locations such as /usr, it becomes non-trivial to deal with files such as /etc/shells which are really an accumulation of several inputs though.
Would it be possible to support a vendor location (e.g. /usr/etc/shells.d/ or /usr/lib/shells.d/), which is trumped by /etc/shells and may contain drop-in files?
That way providers of shell packages on distributions can just add their respective files containing the shell path(s) and pam_shells would just consume them from that vendor location.

The upside to this is, that a system can be run with a removed /etc and be bootstrapped from the vendor location and that this would simplify the packaging of shells on downstream distributions. I'm aware, that other projects would need to buy in on this as well (e.g. shadow-utils, [util-linux](https://github.com/util-linux/util-linux/issues/1857))

**Comment by ldv-alt** [comment]
Some people also suggested in the past that packages could maintain their `shell.d` files and the package management system could maintain `/etc/shells` based on the data in these `shell.d` files, see e.g. https://bugzilla.redhat.com/show_bug.cgi?id=1704972#c3, this is likely to be a simpler approach because it doesn't require coordination between all kinds of projects that might be reading `/etc/shells`.

**Comment by dvzrv** [comment]
> Some people also suggested in the past that packages could maintain their shell.d files and the package management system could maintain /etc/shells based on the data in these shell.d files

The problem with this (current) approach is, that `/etc/shells` is the result of a function of an arbitrary process and all downstreams have to maintain this type of boilerplate. If `/etc/shells` is accumulated from a list of `shell.d` files, this means a process needs to write the file after triggering of a package hook and potentially also on every boot (when thinking of immutable systems). While this can be done (e.g. in the context of systemd using [systemd-generators](https://man.archlinux.org/man/core/systemd/systemd.generator.7.en)), it means even more overhead and a write operation that could be circumvented.
Eventually not requiring an `/etc/shells` file anymore (because one can rely on another well-defined location for drop-in files in the vendor location) seems like a much cleaner approach, because no accumulative action would be required (or would then only be required for backwards compatibility or overrides if needed).

> this is likely to be a simpler approach because it doesn't require coordination between all kinds of projects that might be reading /etc/shells.

While I understand this argument, I don't necessarily agree with it. In the end the only thing (apart from actual implementation of course) that needs to be agreed upon, is from where below `/usr` a given process will optionally read `shell.d` snippet files.
Are there more projects that I've maybe missed, which make use of `/etc/shells`?

**Comment by thkukuk** [comment]
> The upside to this is, that a system can be run with a removed /etc and be bootstrapped from the vendor location and that this would simplify the packaging of shells on downstream distributions. I'm aware, that other projects would need to buy in on this as well (e.g. [shadow](https://github.com/shadow-maint/shadow/issues/587), [util-linux](https://github.com/util-linux/util-linux/issues/1857))

I fully agree that we need a solution for /etc/shells in regards to image based updates as discussed beginning October on the summit in Berlin. Having something like /usr/lib/shells.d, /etc/shells.d and /etc/shells would be great.

But I'm afraid adding support for this to pam_shells is not enough, see getusershell(), setusershell(), endusershell().
This would require changes in glibc, too. 
So we have glibc and all the applications, which don't use the glibc functions like pam_shells.

On the other hand: I couldn't find a common application which uses the glibc functions.

@schubi2 can we not use libeconf for this? Files with only keys, no values? If we have this implemented once for pam_shells it's just Copy&Paste to util-linux and shadow, which are all using libeconf already.
